// @node-wot/browser-bundle@0.9.1/dist/esm/chunk-FEPHGP4Z.js
import {
  __commonJS,
  init_buffer2 as init_buffer,
  require_core
} from "./chunk-WDST5WCP.js";

// ../binding-websockets/dist/ws-client.js
var require_ws_client = __commonJS({
  "../binding-websockets/dist/ws-client.js"(exports) {
    "use strict";
    init_buffer();
    Object.defineProperty(exports, "__esModule", { value: true });
    var core_1 = require_core();
    var { debug, warn } = (0, core_1.createLoggers)("binding-websockets", "ws-client");
    var WebSocketClient = class {
      constructor() {
      }
      toString() {
        return `[WebSocketClient]`;
      }
      readResource(form) {
        return new Promise((resolve, reject) => {
        });
      }
      writeResource(form, content) {
        return new Promise((resolve, reject) => {
        });
      }
      invokeResource(form, content) {
        return new Promise((resolve, reject) => {
        });
      }
      unlinkResource(form) {
        return new Promise((resolve, reject) => {
        });
      }
      subscribeResource(form, next, error, complete) {
        throw new Error("Websocket client does not implement subscribeResource");
      }
      async requestThingDescription(uri) {
        throw new Error("Method not implemented");
      }
      async start() {
      }
      async stop() {
      }
      setSecurity(metadata, credentials) {
        if (metadata === void 0 || !Array.isArray(metadata) || metadata.length === 0) {
          warn("WebSocketClient received empty security metadata");
          return false;
        }
        const security = metadata[0];
        debug(`WebSocketClient using security scheme '${security.scheme}'`);
        return true;
      }
    };
    exports.default = WebSocketClient;
  }
});

// ../binding-websockets/dist/ws-client-factory.js
var require_ws_client_factory = __commonJS({
  "../binding-websockets/dist/ws-client-factory.js"(exports) {
    "use strict";
    init_buffer();
    var __importDefault = exports && exports.__importDefault || function(mod) {
      return mod && mod.__esModule ? mod : { "default": mod };
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    var core_1 = require_core();
    var ws_client_1 = __importDefault(require_ws_client());
    var { debug } = (0, core_1.createLoggers)("binding-websockets", "ws-client-factory");
    var WebSocketClientFactory = class {
      constructor(proxy = null) {
        this.scheme = "ws";
        this.clientSideProxy = null;
        this.clientSideProxy = proxy;
      }
      getClient() {
        debug(`HttpClientFactory creating client for '${this.scheme}'`);
        return new ws_client_1.default();
      }
      init() {
        return true;
      }
      destroy() {
        return true;
      }
    };
    exports.default = WebSocketClientFactory;
  }
});

// ../binding-websockets/dist/wss-client-factory.js
var require_wss_client_factory = __commonJS({
  "../binding-websockets/dist/wss-client-factory.js"(exports) {
    "use strict";
    init_buffer();
    Object.defineProperty(exports, "__esModule", { value: true });
    var WssClientFactory = class {
      constructor() {
        this.scheme = "wss";
      }
      getClient() {
        throw new Error("WssClientFactory for 'wss' is not implemented");
      }
      init() {
        return true;
      }
      destroy() {
        return true;
      }
    };
    exports.default = WssClientFactory;
  }
});

// ../binding-websockets/dist/ws-browser.js
var require_ws_browser = __commonJS({
  "../binding-websockets/dist/ws-browser.js"(exports) {
    init_buffer();
    var __createBinding = exports && exports.__createBinding || (Object.create ? function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      var desc = Object.getOwnPropertyDescriptor(m, k);
      if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
        desc = { enumerable: true, get: function() {
          return m[k];
        } };
      }
      Object.defineProperty(o, k2, desc);
    } : function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    });
    var __exportStar = exports && exports.__exportStar || function(m, exports2) {
      for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports2, p)) __createBinding(exports2, m, p);
    };
    var __importDefault = exports && exports.__importDefault || function(mod) {
      return mod && mod.__esModule ? mod : { "default": mod };
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.WebSocketSecureClientFactory = exports.WebSocketClientFactory = exports.WebSocketClient = void 0;
    var ws_client_1 = require_ws_client();
    Object.defineProperty(exports, "WebSocketClient", { enumerable: true, get: function() {
      return __importDefault(ws_client_1).default;
    } });
    var ws_client_factory_1 = require_ws_client_factory();
    Object.defineProperty(exports, "WebSocketClientFactory", { enumerable: true, get: function() {
      return __importDefault(ws_client_factory_1).default;
    } });
    var wss_client_factory_1 = require_wss_client_factory();
    Object.defineProperty(exports, "WebSocketSecureClientFactory", { enumerable: true, get: function() {
      return __importDefault(wss_client_factory_1).default;
    } });
    __exportStar(require_ws_client(), exports);
    __exportStar(require_ws_client_factory(), exports);
    __exportStar(require_wss_client_factory(), exports);
  }
});

export {
  require_ws_browser
};
