// @node-wot/binding-websockets/dist/ws-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@node-wot/core");
const { debug, warn } = (0, core_1.createLoggers)("binding-websockets", "ws-client");
class WebSocketClient {
    constructor() {
    }
    toString() {
        return `[WebSocketClient]`;
    }
    readResource(form) {
        return new Promise((resolve, reject) => {
        });
    }
    writeResource(form, content) {
        return new Promise((resolve, reject) => {
        });
    }
    invokeResource(form, content) {
        return new Promise((resolve, reject) => {
        });
    }
    unlinkResource(form) {
        return new Promise((resolve, reject) => {
        });
    }
    subscribeResource(form, next, error, complete) {
        throw new Error("Websocket client does not implement subscribeResource");
    }
    async requestThingDescription(uri) {
        throw new Error("Method not implemented");
    }
    async start() {
    }
    async stop() {
    }
    setSecurity(metadata, credentials) {
        if (metadata === undefined || !Array.isArray(metadata) || metadata.length === 0) {
            warn("WebSocketClient received empty security metadata");
            return false;
        }
        const security = metadata[0];
        debug(`WebSocketClient using security scheme '${security.scheme}'`);
        return true;
    }
}
exports.default = WebSocketClient;
