// @node-wot/binding-websockets@0.9.1/dist/ws-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http = __importStar(require("http"));
const https = __importStar(require("https"));
const url = __importStar(require("url"));
const fs = __importStar(require("fs"));
const WebSocket = __importStar(require("ws"));
const core_1 = require("@node-wot/core");
const binding_http_1 = require("@node-wot/binding-http");
const slugify_1 = __importDefault(require("slugify"));
const { debug, info, error } = (0, core_1.createLoggers)("binding-websockets", "ws-server");
class WebSocketServer {
    constructor(serverOrConfig = {}) {
        this.PROPERTY_DIR = "properties";
        this.ACTION_DIR = "actions";
        this.EVENT_DIR = "events";
        this.port = 8081;
        this.address = undefined;
        this.ownServer = true;
        this.thingNames = new Set();
        this.thingPaths = new Map();
        this.socketServers = {};
        if (serverOrConfig instanceof binding_http_1.HttpServer && typeof serverOrConfig.getServer === "function") {
            this.ownServer = false;
            this.httpServer = serverOrConfig.getServer();
            this.port = serverOrConfig.getPort();
            this.scheme = serverOrConfig.scheme === "https" ? "wss" : "ws";
        }
        else if (typeof serverOrConfig === "object") {
            const config = serverOrConfig;
            if (config.port !== undefined) {
                this.port = config.port;
            }
            if (config.address !== undefined) {
                this.address = config.address;
            }
            if (config.serverKey != null && config.serverCert != null) {
                const options = {
                    key: fs.readFileSync(config.serverKey),
                    cert: fs.readFileSync(config.serverCert),
                };
                this.scheme = "wss";
                this.httpServer = https.createServer(options);
            }
            else {
                this.scheme = "ws";
                this.httpServer = http.createServer();
            }
        }
        else {
            throw new Error(`WebSocketServer constructor argument must be HttpServer, HttpConfig, or undefined`);
        }
    }
    start(servient) {
        debug(`WebSocketServer starting on ${this.address !== undefined ? this.address + " " : ""}port ${this.port}`);
        return new Promise((resolve, reject) => {
            this.httpServer.on("upgrade", (request, socket, head) => {
                const pathname = new url.URL(request.url ?? "", `${this.scheme}://${request.headers.host}`).pathname;
                const socketServer = this.socketServers[pathname];
                if (socketServer != null) {
                    socketServer.handleUpgrade(request, socket, head, (ws) => {
                        socketServer.emit("connection", ws, request);
                    });
                }
                else {
                    socket.destroy();
                }
            });
            if (this.ownServer) {
                this.httpServer.once("error", (err) => {
                    reject(err);
                });
                this.httpServer.once("listening", () => {
                    this.httpServer.on("error", (err) => {
                        error(`WebSocketServer on port ${this.port} failed: ${err.message}`);
                    });
                    resolve();
                });
                this.httpServer.listen(this.port, this.address);
            }
            else {
                resolve();
            }
        });
    }
    stop() {
        debug(`WebSocketServer stopping on port ${this.port}`);
        return new Promise((resolve, reject) => {
            for (const socketServer of Object.values(this.socketServers)) {
                socketServer.close();
            }
            if (this.ownServer) {
                debug("WebSocketServer stopping own HTTP server");
                this.httpServer.once("error", (err) => {
                    reject(err);
                });
                this.httpServer.once("close", () => {
                    resolve();
                });
                this.httpServer.close();
            }
        });
    }
    getPort() {
        if (this.httpServer.address() != null && typeof this.httpServer.address() === "object") {
            return this.httpServer.address().port;
        }
        else {
            return -1;
        }
    }
    expose(thing) {
        let urlPath = (0, slugify_1.default)(thing.title, { lower: true });
        if (this.thingNames.has(urlPath)) {
            let uniqueUrlPath;
            let nameClashCnt = 2;
            do {
                uniqueUrlPath = urlPath + "_" + nameClashCnt++;
            } while (this.thingNames.has(uniqueUrlPath));
            urlPath = uniqueUrlPath;
        }
        if (this.getPort() !== -1) {
            debug(`WebSocketServer on port ${this.getPort()} exposes '${thing.title}' as unique '/${urlPath}/*'`);
            this.thingNames.add(urlPath);
            this.thingPaths.set(thing.id, urlPath);
            for (const [propertyName, property] of Object.entries(thing.properties)) {
                const path = "/" +
                    encodeURIComponent(urlPath) +
                    "/" +
                    this.PROPERTY_DIR +
                    "/" +
                    encodeURIComponent(propertyName);
                for (const address of core_1.Helpers.getAddresses()) {
                    const href = `${this.scheme}://${address}:${this.getPort()}${path}`;
                    const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
                    const ops = [];
                    const writeOnly = property.writeOnly ?? false;
                    const readOnly = property.readOnly ?? false;
                    if (!writeOnly) {
                        ops.push("readproperty", "observeproperty", "unobserveproperty");
                    }
                    if (!readOnly) {
                        ops.push("writeproperty");
                    }
                    form.op = ops;
                    thing.properties[propertyName].forms.push(form);
                    debug(`WebSocketServer on port ${this.getPort()} assigns '${href}' to Property '${propertyName}'`);
                }
                debug(`WebSocketServer on port ${this.getPort()} adding socketServer for '${path}'`);
                this.socketServers[path] = new WebSocket.Server({ noServer: true });
                this.socketServers[path].on("connection", (ws, req) => {
                    debug(`WebSocketServer on port ${this.getPort()} received connection for '${path}' from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
                    const observeListener = async (content) => {
                        debug(`WebSocketServer on port ${this.getPort()} publishing to property '${propertyName}' `);
                        for await (const chunk of content.body) {
                            ws.send(chunk);
                        }
                    };
                    const writeOnly = property.writeOnly ?? false;
                    if (writeOnly) {
                        for (let formIndex = 0; formIndex < thing.properties[propertyName].forms.length; formIndex++) {
                            thing
                                .handleObserveProperty(propertyName, observeListener, { formIndex })
                                .catch((err) => ws.close(-1, err.message));
                        }
                    }
                    ws.on("close", () => {
                        for (let formIndex = 0; formIndex < thing.properties[propertyName].forms.length; formIndex++) {
                            thing.handleUnobserveProperty(propertyName, observeListener, { formIndex });
                        }
                        debug(`WebSocketServer on port ${this.getPort()} closed connection for '${path}' from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
                    });
                });
            }
            for (const [actionName, action] of Object.entries(thing.actions)) {
                const path = "/" + encodeURIComponent(urlPath) + "/" + this.ACTION_DIR + "/" + encodeURIComponent(actionName);
                for (const address of core_1.Helpers.getAddresses()) {
                    const href = `${this.scheme}://${address}:${this.getPort()}${path}`;
                    const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
                    form.op = ["invokeaction"];
                    action.forms.push(form);
                    debug(`WebSocketServer on port ${this.getPort()} assigns '${href}' to Action '${actionName}'`);
                }
            }
            for (const [eventName, event] of Object.entries(thing.events)) {
                const path = "/" + encodeURIComponent(urlPath) + "/" + this.EVENT_DIR + "/" + encodeURIComponent(eventName);
                for (const address of core_1.Helpers.getAddresses()) {
                    const href = `${this.scheme}://${address}:${this.getPort()}${path}`;
                    const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
                    form.op = "subscribeevent";
                    event.forms.push(form);
                    debug(`WebSocketServer on port ${this.getPort()} assigns '${href}' to Event '${eventName}'`);
                }
                debug(`WebSocketServer on port ${this.getPort()} adding socketServer for '${path}'`);
                this.socketServers[path] = new WebSocket.Server({ noServer: true });
                this.socketServers[path].on("connection", (ws, req) => {
                    debug(`WebSocketServer on port ${this.getPort()} received connection for '${path}' from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
                    const eventListener = async (content) => {
                        for await (const chunk of content.body) {
                            ws.send(chunk);
                        }
                    };
                    for (let formIndex = 0; formIndex < event.forms.length; formIndex++) {
                        thing
                            .handleSubscribeEvent(eventName, eventListener, { formIndex })
                            .catch((err) => ws.close(-1, err.message));
                    }
                    ws.on("close", () => {
                        for (let formIndex = 0; formIndex < event.forms.length; formIndex++) {
                            thing.handleUnsubscribeEvent(eventName, eventListener, { formIndex });
                        }
                        debug(`WebSocketServer on port ${this.getPort()} closed connection for '${path}' from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
                    });
                });
            }
        }
        return new Promise((resolve, reject) => {
            resolve();
        });
    }
    destroy(thingId) {
        debug(`WebSocketServer on port ${this.getPort()} destroying thingId '${thingId}'`);
        return new Promise((resolve, reject) => {
            let removedThing = false;
            for (const name of Array.from(this.thingPaths.keys())) {
                const thingPath = this.thingPaths.get(name);
                removedThing = this.thingNames.delete(thingPath);
            }
            if (removedThing) {
                info(`WebSocketServer successfully destroyed '${thingId}'`);
            }
            else {
                info(`WebSocketServer failed to destroy thing with thingId '${thingId}'`);
            }
            resolve(removedThing !== undefined);
        });
    }
}
exports.default = WebSocketServer;
